src/noise/rate_limiter.rs

```diff
@@ -0,0 +1,204 @@
+use super::make_array;
+use crate::crypto::blake2s::*;
+use crate::crypto::chacha20poly1305::ChaCha20Poly1305;
+use crate::crypto::x25519::X25519SecretKey;
+
+use crate::noise::handshake::{LABEL_COOKIE, LABEL_MAC1};
+use crate::noise::*;
+use std::net::IpAddr;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::Instant;
+
+const COOKIE_REFRESH: u64 = 128; // Use 128 and not 120 so the compiler can optimize out the division
+const COOKIE_SIZE: usize = 16;
+const COOKIE_NONCE_SIZE: usize = 24;
+
+const RESET_PERIOD: u64 = 1; // How often should reset count in seconds
+
+type Cookie = [u8; COOKIE_SIZE];
+
+// There are two places where WireGuard requires "randomness" for cookies
+// * The 24 byte nonce in the cookie massage - here the only goal is to avoid nonce reuse
+// * A secret value that changes every two minutes
+// Because the main goal of the cookie is simply for a party to prove ownership of an IP address
+// we can relax the randomness definition a bit, in order to avoid locking, because using less
+// resources is the main goal of any DoS prevention mechanism.
+// In order to avoid locking and calls to rand we derive pseudo random values using the AEAD and
+// some counters.
+pub struct RateLimiter {
+    nonce_key: [u8; 32],  // The key we use to derive the nonce
+    secret_key: [u8; 32], // The key we use to derive the secret
+    start_time: Instant,
+    nonce_ctr: AtomicU64, // A single 64bit counter should suffice for many years
+    mac1_key: [u8; 32],
+    cookie_key: [u8; 32],
+    limit: u64,
+    count: AtomicU64,                 // The counter since last reset
+    last_reset: spin::Mutex<Instant>, // The time last reset was performed on this rate limiter
+}
+
+impl RateLimiter {
+    pub fn new(public_key: &X25519PublicKey, limit: u64) -> Self {
+        RateLimiter {
+            nonce_key: RateLimiter::rand_bytes(), // A hack to get 32 random bytes
+            secret_key: RateLimiter::rand_bytes(),
+            start_time: Instant::now(),
+            nonce_ctr: AtomicU64::new(0),
+            mac1_key: Blake2s::new_hash()
+                .hash(LABEL_MAC1)
+                .hash(public_key.as_bytes())
+                .finalize(),
+            cookie_key: Blake2s::new_hash()
+                .hash(LABEL_COOKIE)
+                .hash(public_key.as_bytes())
+                .finalize(),
+            limit,
+            count: AtomicU64::new(0),
+            last_reset: spin::Mutex::new(Instant::now()),
+        }
+    }
+
+    fn rand_bytes() -> [u8; 32] {
+        make_array(X25519SecretKey::new().as_bytes())
```

Comment:
...but this is a hack. Also, X25519 private keys do not have 256 bits of entropy, but 251 bits. Should still be plenty though.

Reply:
The secret key has 256 bits of memory in that case. The mask is applied in the x25519_shared_key function later on.